Broke HttpHandlerSelector into multiple classes.

td-extreme opened the pull request:
Made an abstract HttpHandleRequest class, that has methods for creating default headers, ie server, date.

Created a HandleGetRequest class that extends HttpHandleRequest class.

td-extreme committed:
Broke HttpHandlerSelector into multiple classes.

dadarek commented on src/main/java/HandleGetRequest.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.*;
+import java.io.IOException;
+
+public class HandleGetRequest extends HttpHandleRequest {
```
What's your reasoning for the inheritance?

td-extreme replied:
I guess I was thinking ahead. I was planning on making each type of request. i.e. GET , POST, PUT have its own corresponding class to handle that type of request. However there were some things that would be universal that are in the base class, such as the headers for the server and the date.

dadarek replied:
I see. This was a prep for the next story?

Do it.

dadarek commented on src/main/java/HandleGetRequest.java:
```diff
@@ -0,0 +1,74 @@
+      HashMap<String, String> headers = buildHeaders(body, path);
+      return new HttpResponse(CODE_200, body, headers);
+    }
+    catch (IOException e) {
```
Nitpick -- wrap only that which can throw, and no more.

td-extreme replied:
roger!

dadarek commented on src/main/java/HandleGetRequest.java:
```diff
@@ -0,0 +1,74 @@
+    return Integer.toString(body.length);
+  }
+
+  private String contentType(String fileName) {
```
yuck!

td-extreme replied:
Yuck to the method name??

dadarek replied:
no. the slew of `else if`'s

td-extreme replied:
Gotcha. I could probably make a HashMap<String, String> with file extensions as key and the content type as the value. That's the first thing that comes to mind to me as to how to clean this up. Would that be a good idea, or would you suggest something else that I might consider?

dadarek replied:
HashMap is what I was thinking.

dadarek commented on src/main/java/HttpHandlerSelector.java:
```diff
@@ -1,114 +1,21 @@
-      return "text/html";
+      return getRequest.generateResponse(request);
     }
+    byte[] body = new byte[1];
```
what's the reason for the 1?

td-extreme replied:
This was to make an empty byte array. But, I suppose it should have some actual HTML code for the body that states the file or path could not be found.

dadarek commented on src/main/java/HttpHandlerSelector.java:
```diff
@@ -1,114 +1,21 @@
     }
+    byte[] body = new byte[1];
+    HashMap<String, String> headers = new HashMap<String, String>();
+    return new HttpResponse("HTTP/1.1 404 Not Found", body, headers);
```
Duplicating the 404 response?

td-extreme replied:
Yeah, I will work on correcting this duplication now.

td-extreme replied:
Actually, this really shouldn't be here. I think the problem is that I currently allow any type of request to be valid not just GET, so this was here as a catch for a non GET request.

dadarek commented on src/test/java/HandleGetRequestTest.java:
```diff
@@ -2,9 +2,9 @@
+public class HandleGetRequestTest extends junit.framework.TestCase {
 
-  HttpHandlerSelector handler;
+  HandleGetRequest getRequest;
```
This is confusing. There is a `getRequest`, but there is also a `request` which appears to always be a GET request ... you see what I'm saying?

td-extreme replied:
You mean the name of the instance of the class is confusing, and should probably be named the same as the class, handleGetRequest??

dadarek replied:
No. What's confusing is something like line 25 and 26.

Line 25 creates a new HttpRequest that's really a GET request, and then line 26 uses something that happens to be called `getRequest` to generate a response.

I think what's contributing to the confusion are class names that are actually verbs (i.e. `HandleGetRequest`).

Typically, classes are nouns and methods/functions are verbs.

td-extreme replied:
So the `HandleGetRequest` class is kind of like a Response Factory that builds responses from GET requests. Would a name like `ResponseFactoryGetRequest` make more sense?

If so I would also change the name of the base class from `HttpHandleRequest` to `ResponseFactory`

dadarek replied:
`ResponseFactory` sounds like a good name, `ResponseFactoryGetRequest` not so much. `GetRequestResponseFactory` probably makes more sense.

_However_, factory feels like the wrong abstraction here in general. This thing doesn't feel like it's a factory. It really feels more of a generic "request handler". It takes a request and gives you a response.

Factories are more about hiding the details of how to build or construct an object when that object can be built in different ways, or when we want to encapsulate the details of "how" to build it from the client of that object.

You would use a factory to build a service, or a "thing doer" of some sort that you later use to "do this thing". What's going on here is you're using a class, asking it to "do something or give me something" (based on the Request that's passed in) and what you're getting back is a data object that you _don't_ later use to do something on it, but rather retrieve some fields for presentation.

Perhaps, leave the names as they are for now. Maybe as you build up the next set of requirements a better name will come to our heads.

dadarek commented on src/test/java/HandleGetRequestTest.java:
```diff
@@ -18,48 +18,48 @@
   public void testContentTypeIsTextHtmlForTxtFile() {
     request = new HttpRequest("GET /something.txt HTTP/1.1");
-    response = handler.generateResponse(request);
+    response = getRequest.generateResponse(request);
```
Hmmm ... it looks like this test and the next 3 test are testing some sort of PathToContentTypeMapper thing?

td-extreme replied:
Yeah, they are testing the nasty block of if else above that I am changing to a hashmap. 
 Is it not that clear? Like are you suggesting that the test name be something like `testPathExtenstionTxtMapsToTypeHtmlText`?

dadarek replied:
I was suggesting that maybe a class can be pulled out for this, but in retrospect, maybe it would be overkill.

Not sure.

I guess just leave it for now.

td-extreme replied:
Oh, okay. gotcha.

td-extreme committed:
Made changes based on discussion with Darek

Changed contentType method in HttpHandleRequest method to use a HashMap
instead of a bunch of if else blocks.

Changed try block to only contain the code that could throw exception.

Changed generateResponse method in HttpHandlerSelector to return a valid
request stating that the current method isn’t supported yet for non get
methods. Added a comment to remove this once the support for those other
methods is added, as this will be come dead code.

dadarek commented:
LGTM.

:shipit: unless @mikeknep has something

mikeknep commented on src/main/java/HttpHandleRequest.java:
```diff
@@ -0,0 +1,34 @@
+    HashMap<String, String> rtnHeaders = standardHeaders;
+    rtnHeaders.put("Date", getDateTime());
+    return rtnHeaders;
+  }
```
I find it a little strange to have a private variable `standardHeaders` _and_ a protected method `standardHeaders`. It also seems weird to construct the standard headers in two steps, so to speak (i.e. constructing the hashmap in the class constructor (lines 18-19) and then adding another "static"-ish header in the protected method). Why not something simpler like:

``` java
public HttpHandleRequest() {
  standardHeaders = setupStandardHeaders();
}

protected HashMap<String, String> standardHeaders () {
  standardHeaders = new HashMap<String, String>();
  standardHeaders.put("Server", "My server");
  standardHeaders.put("Date", getDateTime());
  return standardHeaderes;
}
```

td-extreme replied:
Makes sense.

mikeknep commented on src/main/java/HttpHandlerSelector.java:
```diff
@@ -1,114 +1,24 @@
-    }
-    catch (IOException e) {
-      return notFound(request);
+      return getRequest.generateResponse(request);
```
`getRequest` is a little confusing here—it makes me feel like you're calling this on a Request object that happens to be of the GET variety. Probably as simple as renaming to something like `getRequestHandler`

td-extreme replied:
Yeah, Darek and I talked about the confusion with some of the class names. But, aren't really sure yet how to rename them. But, agreed it must be done.

mikeknep commented on src/main/java/HttpHandlerSelector.java:
```diff
@@ -1,114 +1,24 @@
-    this.fileIO = fileIO;
-    standardHeaders = new HashMap<String, String>();
-    standardHeaders.put("Server", "Simple Http Server written in Java");
+  public HttpHandlerSelector(HttpHandleRequest getRequest) {
```
So, if `HttpHandleRequest` is an abstract class, presumably you could be passing in any kind of request handler here, right? Not necessarily a GET request handler? Seems like you'll want a more generic name for the variable getting passed in (`requestHandler`?). Also, shouldn't this approach make the `if` check on `request.method()` below unnecessary?

td-extreme replied:
So, this just made me realize this might be an awful design choice. My original thought was that has I hadn't more types of requests that the constructor would grow to include those as well. So after the POST story it would look like.
`public HttpHandlerSelector(HttpHandleRequest getRequest, HttpHandleRequest postRequest)`
I would then add a HashMap<String, object> with the method types as keys and the instances of HttpHandleRequest as values. This would then remove the `if` check on `request.method` so that it didn't turn into a giant `if`, `else if` block. But, I'm thinking maybe there is a cleaner way to do this. I might need to think on this for a moment.

mikeknep commented on src/main/java/HttpHandlerSelector.java:
```diff
@@ -1,114 +1,24 @@
-      return "text/html";
-    }
+    // TODO :: Remove this after adding classes to handle other types of request methods
+    String responseMessage = "This server currenlty doesn't support request type " + request.method() + "\r\n";
```
Should this whole block be in an `else` clause?

td-extreme replied:
yes, it probably should. I suppose since I knew this would get changed later I didn't bother to do it properly. Which is a bad habit I need to fix. As I'm sure there will be moments in production where code that you might assume will get changed later, won't actually get changed.

td-extreme committed:
Made changes based on Mike’s feedback.

Cleaned up the standardHeaders in HttpHandleRequest class.

Changed the constructor for HttpHandlerSelector so that it takes a
HandleGetRequest object instead of the base class object.

Refactored HttpHandlerSelector to use a HashMap to select the type of
request to process instead of an if else block.

mikeknep commented on src/main/java/HttpServer.java:
```diff
@@ -11,7 +11,8 @@ public static void main(String[] args) throws IOException {
     FileIO fileIO = new FileIO(arguments.getDirectory());
-    HttpHandlerSelector handler = new HttpHandlerSelector(fileIO);
+    HandleGetRequest getRequest = new HandleGetRequest(fileIO);
+    HttpHandlerSelector handler = new HttpHandlerSelector(getRequest);
```
Consider what might happen in the future here. If your `HttpHandlerSelector` needs to be constructed with one of each type of request handler, that means that on every request you're constructing handlers for all types of requests, only to discard all but one of them. Seems a little wasteful, no? I think what you're looking for is something like a Router or a HandlerRegistry...

dadarek replied:
💥

td-extreme replied:
What you're saying makes sense, especially when the time comes to make this server multithreaded. 
I will have to do some research on this either sometime this evening or tomorrow morning. Might be something to bring up at during the apprentice standup, to see if anyone has some good resources that might help.

dadarek commented:
Ditto here. Let's cleanup and merge or close without merge.

td-extreme committed:
Finished organizing classes.

Moved logic for selecting path into HandlerRouter.

HandlerSelector now just uses a method off of HandlerRouter to get the
correct handler.

Removed abstract base class HttpHandleRequest. Handlers no longer use
inheritance, and instead implement the Ihandler interface.

Removed a standard headers builder and instead have a
ResponseHeadersForContent class for building specific get headers.

td-extreme committed:
Removed dead code in test.

td-extreme committed:
Changed contentType to text/plain for .txt files

td-extreme committed:
Updated tests.

td-extreme committed:
Fixed typo

dadarek commented:
Ship it if you're ready.
